### PR TITLE
action: Allow long lines if non-alphabetic

### PR DESCRIPTION
**Pull request by jodh-intel**
Overly long commit lines are annoying. But sometimes, we need to be able to force the use of long lines (for example to reference a URL).

Ironically, I can't refer to the URL that explains this because of ... the long line check! Hence:

```sh
$ cat <<EOT | tr -d '\n'; echo
See: https://github.com/kata-containers/tests/tree/master/
cmd/checkcommits#handling-long-lines
EOT
```

Fixes: #687.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Review by fidencio — APPROVED** [review]
lgtm!

I have the feeling we may end up finding more exceptions in the future, but I'm glad to know that tweaking this is a possibility!

**Comment by c3d** [comment]
@jodh-intel now the body line check failed 😃

File: .github/workflows/commit-message-check.yaml [hunk 1]
```diff
@@ -48,7 +48,7 @@ jobs:
       uses: tim-actions/commit-message-checker-with-regex@v0.3.1
       with:
         commits: ${{ steps.get-pr-commits.outputs.commits }}
-        pattern: '^.+(\n.{0,72})*$|^.+\n\s*[^a-zA-Z\s\n]|^.+\n\S+$'
+        pattern: '^[a-zA-Z]+(\n.{0,72})*$|^[a-zA-Z]+\n\s*[^a-zA-Z\s\n]|^[a-zA-Z]+\n\S+$'
```
Comment (c3d) [inline .github/workflows/commit-message-check.yaml:51]: I don't understand the pattern. From https://github.com/GsActions/commit-message-checker, it looks like the pattern includes the title of the commit. So you are testing `[a-zA-Z]` on that title, and asking it to consist of only alpha characters before the `\n`? It looks wrong.

I would try the following pattern:

```
^[a-zA-Z].{0,71}(\n([a-zA-Z].{0,71}|[^a-zA-Z\n].*|))+$
```

(Tested using https://regex101.com, seems to give the correct results)

**Comment by bergwolf** [comment]
Not an expression expert at all. But I would love to say that we should definitely allow extremely long commit message lines. e.g., sometimes we need to paste some backtrace logs to the commit message. It would be frustrating if we have to manually wrap the logs.

File: .github/workflows/commit-message-check.yaml (hunk 1 above)
Comment (Tim-Zhang) [reply .github/workflows/commit-message-check.yaml:51]: The `.+` before the `\n` is used to match the title, so we can't replace it with [a-zA-Z] for titles always include colons and numbers

**Comment by Tim-Zhang** [comment]
@jodh-intel  There are some problems with the regex and  I think we have already support long lines. Did you have a case that failed the test?

File: .github/workflows/commit-message-check.yaml (hunk 1 above)
Comment (Tim-Zhang) [reply .github/workflows/commit-message-check.yaml:51]: The second part of the regex `^.+\n\s*[^a-zA-Z\s\n]` is for the long lines support, `\s*[^a-zA-Z\s\n]` will match non-alphabetic.

**Comment by jodh-intel** [comment]
Thanks @c3d, @Tim-Zhang - branch updated and tested locally. Note that I've also fixed the SoB handling to match https://github.com/kata-containers/tests/tree/master/cmd/checkcommits#handling-long-lines and added comments. ptal.

@Tim-Zhang - one other point - I thought workflow files were supposed to live in the `master` branch, so they applied to all branches? If so, shouldn't this file be moved? I *really* wish the actions could live in https://github.com/kata-containers/.github instead tbh ;(

File: .github/workflows/commit-message-check.yaml [hunk 2]
```diff
@@ -48,7 +48,25 @@ jobs:
...
+        #   This allows stack traces, log files snippets, emails, long URLs,
+        #   etc to be specified. Some of these naturally "work" as they start
+        #   with numeric timestamps or addresses. Emails can but quoted using
+        #   the normal ">" character, markdown bullets ("-", "*") are also
+        #   useful for lists of URLs, but it is always possible to override
+        #   the check by simply space indenting the content you need to add.
+        #
+        # - A SoB comment can be any length (as it is unreasonable to penalise
+        #   people with long names/email addresses :)
+        pattern: '^.+(\n([a-zA-Z].{0,71}|[^a-zA-Z\n].*|Signed-off-by:.*|))+$'
```
Comment (Tim-Zhang) [inline .github/workflows/commit-message-check.yaml:69]: Why did you delete the single word line support with the last part of original regex `^.+\n\S+$`

Comment (Tim-Zhang) [reply .github/workflows/commit-message-check.yaml:69]: FYI: https://github.com/kata-containers/tests/blob/master/cmd/checkcommits/checkcommits.go#L201

Comment (Tim-Zhang) [reply .github/workflows/commit-message-check.yaml:69]: The other parts looks good to me

**Comment by Tim-Zhang** [comment]
@jodh-intel I have leave a comment about the regex. 

After my test, the workflow files in .github does not seem to work about `These files apply to all other Kata Containers repositories that do not have their own versions in their .github/ directory`

File: .github/workflows/commit-message-check.yaml (hunk 2 above)
Comment (Tim-Zhang) [reply .github/workflows/commit-message-check.yaml:69]: What about this one: `^.+(\n\s*([a-zA-Z].{0,71}|[^a-zA-Z\s\n].*|Signed-off-by:.*|\S+\s*))+$`
1. Support single word line
2. Because there will be spaces in the beginning of the line, add `\s*` after the `\n`
3. long lines should start with non-alphabetic character but not a space so I add \s in the part `[^a-zA-Z\n]`

Comment (jodh-intel) [reply .github/workflows/commit-message-check.yaml:69]: > Why did you delete the single word line support with the last part of original regex

It's redundant fwics: `[a-zA-Z].{0,71}` matches any number of words due to the dot.

> long lines should start with non-alphabetic character but not a space

We've always allowed long lines if they start with a space (see https://github.com/kata-containers/tests/tree/master/cmd/checkcommits#handling-long-lines), but that should be the exception. "Normal" lines should *not* start with a space.

**Comment by jodh-intel** [comment]
> After my test, the workflow files in .github repo does not seem to work about the point These files apply to all other Kata
> Containers repositories that do not have their own versions in their .github/ directory

Hi @Tim-Zhang - I'm not quite clear what you mean here. Putting workflow files in https://github.com/kata-containers/.github/tree/master/.github/ does not work for other Kata repo fwics.

File: .github/workflows/commit-message-check.yaml (hunk 2 above)
Comment (Tim-Zhang) [reply .github/workflows/commit-message-check.yaml]: 1. The dot only match one character, so `[a-zA-Z].{0,71}` can only match a word with length less than 72. But I think we should match words of any length in order to support url.
2. The space problem is my misunderstanding.

So I think the regex could be:
`^.+(\n([a-zA-Z].{0,149}|[^a-zA-Z\n].*|Signed-off-by:.*|\S+\s*))+$`

**Comment by Tim-Zhang** [comment]
> > After my test, the workflow files in .github repo does not seem to work about the point These files apply to all other Kata
> > Containers repositories that do not have their own versions in their .github/ directory
> 
> Hi @Tim-Zhang - I'm not quite clear what you mean here. Putting workflow files in https://github.com/kata-containers/.github/tree/master/.github/ does not work for other Kata repo fwics.

What I mean is the workflow file in .git repo will not automatically apply to all other repos, May be we should copy it manually. 

So it doesn't make sense to move workflow files into https://github.com/kata-containers/.github

**Comment by jodh-intel** [comment]
@Tim-Zhang - I agree! In fact, imho the whole workflow GitHub feature is flawed (atleast for our workflows :) because of this.

The ideal in my mind is that each repo has a very small number of workflow files (maybe two: one for issues and one for PRs), and those files simply call a script in https://github.com/kata-containers/.github/tree/master/scripts to handle *all* checks for issues + PRs. That way, the workflow files won't need to be modified: they are just "stubs" - we just update the scripts in https://github.com/kata-containers/.github/tree/master/scripts when we want to modify the behaviour of these checks for *all* our repos. This is what I did for the [PR porting checks](https://github.com/kata-containers/kata-containers/issues/634). For example: https://github.com/kata-containers/agent/pull/844/files which removes all the hard-coded logic and just calls https://github.com/kata-containers/.github/blob/master/scripts/pr-porting-checks.sh instead.

The problem with that idea is that it would break Javascript actions as they would need to be rewritten using a script.

**Comment by Tim-Zhang** [comment]
> @Tim-Zhang - I agree! In fact, imho the whole workflow GitHub feature is flawed (atleast for our workflows :) because of this.
> 
> The ideal in my mind is that each repo has a very small number of workflow files (maybe two: one for issues and one for PRs), and those files simply call a script in https://github.com/kata-containers/.github/tree/master/scripts to handle _all_ checks for issues + PRs. That way, the workflow files won't need to be modified: they are just "stubs" - we just update the scripts in https://github.com/kata-containers/.github/tree/master/scripts when we want to modify the behaviour of these checks for _all_ our repos. This is what I did for the [PR porting checks](https://github.com/kata-containers/kata-containers/issues/634). For example: https://github.com/kata-containers/agent/pull/844/files which removes all the hard-coded logic and just calls https://github.com/kata-containers/.github/blob/master/scripts/pr-porting-checks.sh instead.
> 
> The problem with that idea is that it would break Javascript actions as they would need to be rewritten using a script.

We can move the regex into the github-action, in this way, the workflow files won't need to be modified too - we just update the actions. 

I shouldn't have used the regex, it is too complicated to maintain. that's not the way how action works, This is just my tricky way to do multiple checkes with a common action.

**Comment by Tim-Zhang** [comment]
@jodh-intel I can transfer the actions what we are using to the kata-containers org. We can control the actions

**Comment by Tim-Zhang** [comment]
@jodh-intel I can write a new github action called `kata-commit-message-check` move all logic about commit-checks into it.

**Comment by jodh-intel** [comment]
@Tim-Zhang - I don't know what's possible with JS/GraphQL, but the cleanest solution might be to have something like this in each repo (and an equivalent one for issues and other events?):

```yaml
on: [ <list-of-all-pr-events> ]
jobs:
  handle-event:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code to allow hub to communicate with the project
        uses: actions/checkout@v2
        path: main

      - name: Checkout tooling
        uses: actions/checkout@v2
        repository: kata-containers/.github
        path: scripts

      - name: Run Kata actions appropriate for this PR event
        run: |
          scripts/scripts/kata-handle-gh-actions.sh <<< "${{ toJson(github) }}"
```

Then, https://github.com/kata-containers/.github/tree/master/scripts/kata-handle-gh-actions.sh could call a set of actions in https://github.com/kata-containers/.github/tree/master/actions/*.sh` for example. And they would all be passed the JSON too so they can decide if they need to simply `exit 0` or actually perform some checks and then `exit [0|1]`.

**Comment by jodh-intel** [comment]
>  I think we have already support long lines

@Tim-Zhang - It isn't working (see https://github.com/kata-containers/kata-containers/pull/795/). That's what this PR is trying to fix.

/cc @c3d.

Re-pushing as Travis seems unhappy about something...

File: .github/workflows/commit-message-check.yaml (hunk 2 above)
Comment (Tim-Zhang) [reply .github/workflows/commit-message-check.yaml]: any updates about the single word line? @jodh-intel

**Comment by Tim-Zhang** [comment]
@jodh-intel I think we should push this pr merge quickly to fix bugs, before that do you mind update the regex to support single word line?

**Comment by jodh-intel** [comment]
Hi @Tim-Zhang - The regex on this PR already supports single word lines (starting with alphabetics) due to `[a-zA-Z].{0,149}` (which matches "a letter followed by zero or more (149) characters (which could include spaces, alphabetics, *etc*)".

**Comment by Tim-Zhang** [comment]
@jodh-intel LGTM

**Review by Tim-Zhang — APPROVED** [review]
